File: 프로그래머스/0/181918. 배열 만들기 4/배열 만들기 4.js
function solution(arr) {
    let stk = [];
    let i = 0;

    while (i < arr.length) {
        if (!stk.length || stk[stk.length - 1] < arr[i]) {
            stk.push(arr[i]);
            i++;
        } else {
            stk.pop();
        }
    }
    
    return stk;
}

File: 프로그래머스/0/181918. 배열 만들기 4/배열 만들기 4.java
import java.util.*;

class Solution {
    public int[] solution(int[] arr) {
        int i = 0;
        List<Integer> stk = new ArrayList<>();
        
        while (i < arr.length) {
            if (stk.isEmpty()) {
                stk.add(arr[i]);
                i++;
            } else {
                int lastIdx = stk.size() - 1;
                
                if (stk.get(lastIdx) < arr[i]) {
                    stk.add(arr[i]);
                    i++;
                } else {
                    stk.remove(lastIdx);
                }
            }
        }
        
        return stk.stream().mapToInt(Integer::intValue).toArray();
    }
}